db/db_follower_test.cc

```diff
@@ -53,6 +253,267 @@ TEST_F(DBFollowerTest, Basic) {
+  ASSERT_OK(Flush());
+  ASSERT_OK(Put("k1", "v4"));
+  ASSERT_OK(Flush());
+
+  ASSERT_OK(dbfull()->TEST_CompactRange(0, nullptr, nullptr, nullptr, true));
+
+  TEST_SYNC_POINT("Follower::WaitForCatchup");
+  ASSERT_EQ(FollowerGet("k1"), "v4");
+
+  SyncPoint::GetInstance()->ClearAllCallBacks();
+  SyncPoint::GetInstance()->DisableProcessing();
+}
+
+// This test validates the same as the previous test, except there is a
+// MANIFEST rollover between the flushes and compaction. The follower
+// does not switch to a new MANIFEST in ReadAndApply. So it would require
+// another round of refresh before catching up.
+TEST_F(DBFollowerTest, RetryCatchupManifestRollover) {
+  Options opts = CurrentOptions();
+  opts.disable_auto_compactions = true;
+  Reopen(opts);
+
+  ASSERT_OK(Put("k1", "v1"));
+  ASSERT_OK(Flush());
+  ASSERT_OK(Put("k1", "v2"));
+  ASSERT_OK(Flush());
+  ASSERT_OK(Put("k1", "v3"));
+  ASSERT_OK(Flush());
+  ASSERT_OK(OpenAsFollower());
+  SyncPoint::GetInstance()->LoadDependency({
+      {"DBImplFollower::TryCatchupWithLeader:Begin1", "Leader::Start"},
+      {"Leader::Flushed", "DBImplFollower::TryCatchupWithLeader:Begin2"},
+      {"VersionEditHandlerPointInTime::MaybeCreateVersion:Begin1",
+       "Leader::Done"},
+      {"DBImpl::BackgroundCallCompaction:PurgedObsoleteFiles",
+       "VersionEditHandlerPointInTime::MaybeCreateVersion:Begin2"},
+      {"DBImplFollower::TryCatchupWithLeader:End",
+       "Follower::WaitForCatchup:1"},
+  });
+  SyncPoint::GetInstance()->EnableProcessing();
+  TEST_SYNC_POINT("Leader::Start");
+  ASSERT_OK(Put("k1", "v4"));
+  ASSERT_OK(Flush());
+
+  // FlushOptions fo;
+  // fo.wait = false;
+  // ASSERT_OK(dbfull()->Flush(fo, dbfull()->DefaultColumnFamily()));
+  TEST_SYNC_POINT("Leader::Flushed");
+  TEST_SYNC_POINT("Leader::Done");
+  // This should trigger a compaction
+  // WaitForCompactOptions wfco;
+  // ASSERT_OK(dbfull()->WaitForCompact(wfco));
+  Reopen(opts);
+  ASSERT_OK(dbfull()->TEST_CompactRange(0, nullptr, nullptr, nullptr, true));
+
+  TEST_SYNC_POINT("Follower::WaitForCatchup:1");
+  SyncPoint::GetInstance()->LoadDependency({
+      {"DBImplFollower::TryCatchupWithLeader:End",
+       "Follower::WaitForCatchup:2"},
+  });
+  TEST_SYNC_POINT("Follower::WaitForCatchup:2");
+  ASSERT_EQ(FollowerGet("k1"), "v4");
+
+  SyncPoint::GetInstance()->ClearAllCallBacks();
+  SyncPoint::GetInstance()->DisableProcessing();
+}
+
+// This test creates 4 L0 files and compacts them. The follower, during catchup,
+// successfully instantiates 4 Versions corresponding to the 4 files (but
+// donesn't install them yet), followed by deleting those 4 and adding a new
+// file from compaction. The test verifies that the 4 L0 files are deleted
+// correctly by the follower.
+TEST_F(DBFollowerTest, IntermediateObsoleteFiles) {
+  Options opts = CurrentOptions();
+  opts.disable_auto_compactions = true;
+  Reopen(opts);
+  ASSERT_OK(OpenAsFollower());
+
+  follower_fs()->BarrierInit(2);
```

Comment:
I didn't quite get what these `Barrier*` functions helped to achieve, would you mind adding a bit clarification?